[All hosts](SSO) Update walkthrough for ASP.NET Core (#3567)
OfficeDev/office-js-docs-pr: docs/develop/create-sso-office-add-ins-aspnet.md

---
title: Create an ASP.NET Office Add-in that uses single sign-on
description: A step-by-step guide for how to create (or convert) an Office Add-in with an ASP.NET backend to use single sign-on (SSO).
ms.date: 07/29/2022
ms.localizationpriority: medium
---

# Create an ASP.NET Office Add-in that uses single sign-on

Users can sign in to Office, and your Office Web Add-in can take advantage of this sign-in process to authorize users to your add-in and to Microsoft Graph without requiring users to sign in a second time. For an overview, see [Enable SSO in an Office Add-in](sso-in-office-add-ins.md).

> This article walks you through the process of enabling single sign-on (SSO) in an add-in. The sample add-in you create has two parts; a task pane that loads in Microsoft Excel, and a middle-tier server that handles calls to Microsoft Graph for the task pane. The middle-tier server is built ASP.NET core and exposes a single REST API, `/api/filenames`, that returns a list of the first 10 file names in the user's OneDrive folder. The task pane uses the `getAccessToken()` method to get an access token for the signed in user to the middle-tier server. The middle-tier server uses the On-Behalf-Of flow (OBO) to exchange the access token for a new one with access to Microsoft Graph. You can extend this pattern to access any Microsoft Graph data. The task pane always calls a middle-tier REST API (passing the access token) when it needs Microsoft Graph services. The middle-tier uses the token obtained via OBO to call Microsoft Graph services and return the results to the task pane.

Author: Makes sense. I updated this.

---

> title: Create an ASP.NET Office Add-in that uses single sign-on

Reviewer: For SEO, consider adding "(SSO)" to title since people use it in Search.

Author: Added.

## Register the add-in with Microsoft identity platform

You need to create an app registration in Azure that represents your middle-tier server. This enables authentication support so that proper access tokens can be issued to the client code in JavaScript. This registration supports both SSO in the client, and fallback authentication using the Microsoft Authentication Library (MSAL).

1. To register your app, navigate to the [Azure portal - App registrations](https://go.microsoft.com/fwlink/?linkid=2083908) page to register your app.

1. Sign in with the **_admin_** credentials to your Microsoft 365 tenancy. For example, MyName@contoso.onmicrosoft.com.

1. Select **New registration**. On the **Register an application** page, set the values as follows.

   - Set **Name** to `Office-Add-in-ASPNET-SSO`.
   - Set **Supported account types** to **Accounts in any organizational directory (Any Azure AD directory - Multitenant) and personal Microsoft accounts (e.g. Skype, Xbox).**.

>    - In the **Redirect URI** section, set the platform to **Single-page application (SPA)** with a redirect URI value of `https://localhost:7283/dialog.html`.

Author: fixed.

   - Choose **Register**.

   > [!NOTE]
   > The SPA application type is only used when the client uses MSAL for fallback authentication.

1. On the **Office-Add-in-ASPNET-SSO** page, copy and save the **Application (client) ID**. You'll use it in later procedures.

   > [!NOTE]
   > This **Application (client) ID** is the "audience" value when other applications, such as the Office client application (e.g., PowerPoint, Word, Excel), seek authorized access to the application. It's also the "client ID" of the application when it seeks authorized access to Microsoft Graph.

> 1. In the leftmost sidebar, select **Authentication** under **Manage**. In the **Implicit grant and hybrid flows** section, select both checkboxes for **Access tokens** and **ID tokens**. The sample uses the Microsoft Authentication Library (MSAL) for fallback authentication when SSO is not available.

Author: fixed

   > [!NOTE]
   > The **\<Resource\>** value is the **Application ID URI** you set when you registered the add-in. The **\<Scopes\>** section is used only to generate a consent dialog box if the add-in is sold through AppSource.

1. Save and close the manifest file.

1. In **Solution Explorer**, choose the **Office-Add-in-ASPNET-SSO-web** project and open the **appsettings.json** file.

1. Replace the placeholder **$app-id-guid$** with the **Application (client) ID** value you saved previously.

1. Replace the placeholder **$client-secret$** with the client secret value you saved previously.

1. Replace the **$publisher-domain$** with the domain name of your app registration. You can find your publisher domain by going to the App registration you created previously. In the leftmost sidebar, select **Manifest** under **Manage**. Search for the **publisherDomain** value in the manifest JSON.

    > [!NOTE]

>     > You can also change the **TenantId** to support single-tenant if you configured your app registration for single-tenant. Replace the **Common** value with the **Application (client) ID** for single-tenant support.

Author: must, and fixed.

1. Save and close the appsettings.json file.

> 1. Open the **wwwroot/js/fallback-msal/authConfig.js** file. This file specifies the configuration for using MSAL for fallback authentication.

Author: fallback code is removed now, but if you did use fallback, you would use both js (client side) and .net (server side).

1. In the **Office-Add-in-ASPNET-SSO-web** project, open the **wwwroot\js\ssoAuthES6.js** file.  It already has code that ensures that Promises are supported, even in Internet Explorer 11, and an `Office.onReady` call to assign a handler to the add-in's only button.

   > [!NOTE]
   > As the name suggests, the ssoAuthES6.js uses JavaScript ES6 syntax because using `async` and `await` best shows the essential simplicity of the SSO API. When the localhost server is started, this file is transpiled to ES5 syntax so that the sample will support Internet Explorer 11.

    A key part of the sample code is the client request. The client request is an object that tracks information about the request for calling REST APIs on the middle-tier server. It's necessary because the client request state needs to be tracked or updated through the following scenarios:

    - SSO retries where the REST API call fails because it needs additional consent. The sample code calls `getAccessToken` with updated authentication options, gets the required user consent, and then calls the REST API again. The goal is to not fail in scenarios where a REST API needs additional consent.

>     - SSO fails and fallback authentication is required. The access token is acquired through MSAL in a pop-up dialog box. The goal is to not fail in this scenario and gracefully fall back to the alternative authentication approach.

Author: fallback steps are removed now.

    The client request object tracks the following data:

    - `authOptions` - [Auth configuration parameters](/javascript/api/office/office.authoptions) for SSO.
    - `authSSO` - true if using SSO, otherwise false.
    - `verb` - REST API verb such as GET, POST...

>     - `accessToken` - The access token to the middle-tier server. The method to obtain this token is different for SSO than fallback auth.

Author: This section is removed now.

            );
            break;
        case 13008:
            // Only seen in Office on the web.
            showMessage(
                "Office is still working on the last operation. When it completes, try this operation again."
            );
            break;
        case 13010:
            // Only seen in Office on the web.
            showMessage(
                "Follow the instructions to change your browser's zone configuration."
            );
            break;
   ```

1. Replace `TODO 8` with the following code. For more information about these errors, see [Troubleshoot SSO in Office Add-ins](troubleshoot-sso-in-office-add-ins.md). For any errors that can't be handled, `true` is returned to the caller. This indicates the caller should switch to using MSAL as fallback auth.

   ```javascript
    default:
            // For all other errors, including 13000, 13003, 13005, 13007, 13012, and 50001, fall back
            // to non-SSO sign-in.
            fallbackRequired = true;
            break;
    }
    return fallbackRequired;
   ```

### Add fallback to MSAL auth

> Fallback authentication uses the MSAL library to sign in the user. The add-in itself is an SPA, and uses an SPA app registration to access the middle-tier server.

Author: yep, this has changed and it is not an SPA.

> [!NOTE]
> There are two versions of the sample.
>
> - The **Begin** folder is a starter project. The UI and other aspects of the add-in that are not directly connected to SSO or authorization are already done. Later sections of this article walk you through the process of completing it.
> - The **Complete** folder contains the same sample with all coding steps from this article completed. To use the completed version, just follow the instructions in this article, but replace "Begin" with "Complete" and skip the sections **Code the client side** and **Code the server side**.

Use the following values for placeholders for the subsequent app registration steps.

| Placeholder           | Value                                           |
|-----------------------|-------------------------------------------------|
| `<add-in-name>`       | **Office-Add-in-ASPNET-SSO**                    |
| `<redirect-platform>` | Leave this blank. The SSO sample does not use a redirect URI. |

> | `<redirect-uri>`      | Leave this blank. The SSO sample does not use a redirect URI. |

Reviewer 2: If these aren't used, why include them?

Author: This is an include file for Node.js and ASP.NET walkthroughs. ASP.NET does not use these. However, let me see if I can make this clearer.

Author: I removed these and updated the include file to indicate which value to use for Node.js vs ASP.NET.

    ```javascript
    else {
        dialogFallback();
    // Results from other errors (other than AADSTS50076) will have an ExceptionMessage property.
    const exceptionMessage = JSON.parse(errorResponse.responseText).ExceptionMessage;
    if (exceptionMessage) {
        // On rare occasions the access token is unexpired when Office validates it,
        // but expires by the time it is sent to Microsoft identity in the OBO flow. Microsoft identity will respond
        // with "The provided value for the 'assertion' is not valid. The assertion has expired."
        // Retry the call of getAccessToken (no more than once). This time Office will return a 
        // new unexpired access token.
        if ((exceptionMessage.includes("AADSTS500133"))
            && (retryGetAccessToken <= 0)) {
            retryGetAccessToken++;
            getUserFileNames();
            return;
        }
        else {
            showMessage("MSAL error from application server: " + JSON.stringify(exceptionMessage));
            return;
        }
    }
    // Default error handling if previous checks didn't apply.
    showMessage(errorResponse.responseJSON.value);
    ```

> 1. Save the file.

Reviewer 2: I think this was accidentally removed.

Author: restored

> The server-side code is an ASP.NET Core server that provides REST APIs for the client to call. For example, the REST API `/api/files` gets a list of filenames from the user's OneDrive folder. Each REST API call requires an access token by the client to ensure the correct client is accessing their data. The access token is exchanged for a Microsoft Graph token through the On-Behalf-Of flow (OBO). The new Microsoft Graph token is cached by the MSAL.NET library for subsequent API calls. It's never sent outside of the server-side code. Microsoft identity documentation refers to this server as the middle-tier server because it is in the middle of the flow from client-side code to Microsoft services. For more information, see [Middle-tier access token request](/azure/active-directory/develop/v2-oauth2-on-behalf-of-flow#middle-tier-access-token-request)

Reviewer 3: Thanks for adding this call out! Important security considerations for our customers! :)

> 1. Open the Visual Studio solution file.

Author: true, updated.

> 1. On the **Build** menu, select **Clean Solution**. When it finishes, open the **Build** menu again and select **Build Solution**.

Reviewer 2: Reinstalling VS fixed this.

> 1. In the Office application, on the **Home** ribbon, select the **Show Add-in** in the **SSO ASP.NET** group to open the task pane add-in.

Reviewer: ```suggestion
1. In the Office application, select the **Show Add-in** in the **SSO ASP.NET** group to open the task pane add-in.
```
We don't use "on the ribbon" anymore.

> 1. Click the **Get OneDrive File Names** button. If you are logged into Office with either a Microsoft 365 Education or work account, or a Microsoft account, and SSO is working as expected, the first 10 file and folder names in your OneDrive for Business are displayed on the task pane. If you are not logged in, or you are in a scenario that does not support SSO, or SSO is not working for any reason, you will be prompted to sign in. After you sign in, the file and folder names appear.

Reviewer: ```suggestion
1. Select **Get OneDrive File Names**. If you're logged into Office with either a Microsoft 365 Education or work account, or a Microsoft account, and SSO is working as expected, the first 10 file and folder names in your OneDrive for Business are displayed on the task pane. If you are not logged in, or you are in a scenario that does not support SSO, or SSO is not working for any reason, you will be prompted to sign in. After you sign in, the file and folder names appear.
```
We don't say "click the _blah_ button" anymore, either.

> When you are ready to deploy to a staging or production server, be sure to update the following areas in the project solution.

Reviewer: ```suggestion
When you're ready to deploy to a staging or production server, be sure to update the following areas in the project solution.
```
